### PR TITLE
Reproduce NA2Q value decomposition on cooperative MARL

### LINKED ISSUE #57: [Notebook] Reproduce NA2Q on one cooperative MARL task
Parent roadmap: #60  
Library dependencies: #52, #53, and #62

## Paper

[NA2Q: Neural Attention Additive Model for Interpretable Multi-Agent Q-Learning](https://proceedings.mlr.press/v202/liu23be.html)  
Reference code: [zichuan-liu/NA2Q](https://github.com/zichuan-liu/NA2Q)

## Outcome

Deliver `docs/source/reproductions/na2q-value-decomposition.ipynb`, a dedicated scientific reproduction notebook separate from tutorials.

On one fixed cooperative task, the notebook should:

1. reproduce the neural additive/coalition value decomposition;
2. expose identity semantics and local semantic masks;
3. verify joint-value or induced-policy computation against the reference; and
4. compare task and interpretability metrics with declared baselines.

## Notebook requirements

- Pin the NA2Q revision, environment/scenario, checkpoint or training config, seeds, and evaluation protocol.
- Map PyMARL tensors/modules to TensorDict keys, agent/coalition axes, value roles, masks, and named reductions using #62.
- Establish deterministic forward parity for a saved reference batch.
- Preserve per-agent and per-coalition outputs until an explicit named reduction.
- Report return/win-rate and declared interpretability metrics with uncertainty and frozen controls.
- Record external inputs and full-result artifacts through #53.
- Include a bounded smoke path clearly marked as non-scientific.

## Acceptance criteria

- [ ] The dedicated notebook renders from the Reproductions gallery introduced by #52.
- [ ] Additive/coalition outputs and masks satisfy a reference fixture or declared invariants.
- [ ] Parameter-shared modules do not masquerade as semantic agents or coalitions.
- [ ] Reference differences and unsupported PyMARL behavior are explicit.
- [ ] One scenario is not used to claim all-task superiority or “human-like interpretability.”

## Non-goals

This issue does not port all of PyMARL or add NA2Q to XDRL core.

### PR DESCRIPTION
**Pull request by Xmaster6y**
## Summary

- add a dedicated, executed NA2Q value-decomposition reproduction notebook for the pinned `lbf-4-2` protocol
- map individual values, unary/pair coalitions, identity semantics, local masks, attention, state bias, and named reductions into XDRL TensorDict contracts
- verify saved-fixture and instrumented parity, additive reconstruction, monotonicity, induced-policy consistency, typed artifact provenance, uncertainty reporting, and frozen controls
- fail closed when exact paper checkpoints/results are unavailable and keep smoke evidence separate from scientific claims
- expose the notebook from the Reproductions gallery

## Validation

- `.venv/bin/jupyter-nbconvert --to notebook --execute --inplace docs/source/reproductions/na2q-value-decomposition.ipynb --ExecutePreprocessor.timeout=60`
- `.venv/bin/pre-commit run --all-files`
- `.venv/bin/pytest tests --cov=src --cov-report=term-missing --cov-fail-under=50 -q` (284 passed; 95.97% coverage)
- `PATH="/Users/xmaster/Work/xdrl/.venv/bin:$PATH" make html SPHINXOPTS="-W --keep-going"` from `docs/`

closes #57

### CODE REVIEW

File: docs/source/reproductions/na2q-value-decomposition.ipynb [hunk 1]
```diff
@@ -0,0 +1,966 @@
...
+    "            selected = individual_value.squeeze(-1).unsqueeze(1) * coalition_mask_value\n",
+    "            sizes = coalition_mask_value.sum(-1)\n",
+    "            total = selected.sum(-1)\n",
+    "            product = torch.where(\n",
+    "                sizes == 2,\n",
+    "                torch.where(coalition_mask_value.bool(), individual_value.squeeze(-1).unsqueeze(1), 1).prod(-1),\n",
+    "                torch.zeros_like(total),\n",
+    "            )\n",
+    "            raw = torch.where(sizes == 1, 0.4 * total + 0.13, 0.25 * total + 0.1 * product + 0.08).unsqueeze(-1)\n",
+    "            state_signal = context[:, :1]\n",
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/na2q-value-decomposition.ipynb]: Fixed in e5c7f6c. Smoke attention now projects the declared `identity_semantics` tensor into per-agent identity codes and uses those codes under each coalition mask; hard-coded member IDs were removed.

File: docs/source/reproductions/na2q-value-decomposition.ipynb [hunk 2]
```diff
@@ -0,0 +1,966 @@
...
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "def mean_ci(values):\n",
+    "    values = torch.as_tensor(values, dtype=torch.float).flatten()\n",
+    "    generator = torch.Generator().manual_seed(991)\n",
+    "    draws = torch.randint(len(values), (2000, len(values)), generator=generator)\n",
+    "    bootstrap = values[draws].mean(-1)\n",
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/na2q-value-decomposition.ipynb]: Fixed in e5c7f6c. `task_return_ci` now requires `(5, 32)`, averages episodes within each seed, and bootstraps the five seed-level means, so task intervals report `n=5`.

File: docs/source/reproductions/na2q-value-decomposition.ipynb [hunk 3]
```diff
@@ -0,0 +1,966 @@
...
+    "    out_keys=[RAW_KEY, ATTENTION_KEY, BIAS_KEY, KEYS.coalition_contribution, KEYS.joint_value],\n",
+    ")\n",
+    "native = RuntimeInteractionContext(contract, mixer, data)(data.clone())\n",
+    "\n",
+    "if MODE == \"smoke\":\n",
+    "    torch.testing.assert_close(native[RAW_KEY][0, :, 0], bundle[\"reference_raw_shape_first\"], rtol=0, atol=1e-7)\n",
+    "    torch.testing.assert_close(native[ATTENTION_KEY][0, :, 0], bundle[\"reference_attention_first\"], rtol=0, atol=1e-7)\n",
+    "    torch.testing.assert_close(native[KEYS.joint_value][0, 0], bundle[\"reference_joint_first\"], rtol=0, atol=1e-7)\n",
+    "else:\n",
+    "    torch.testing.assert_close(native[KEYS.joint_value], bundle[\"reference_joint_value\"], rtol=1e-6, atol=1e-7)\n",
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/na2q-value-decomposition.ipynb]: Fixed in e5c7f6c. The paper-mode assertion is now a recorded export-reconstruction gate; failed comparisons reach the verdict and retain the full evidence and typed provenance artifacts.

[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/na2q-value-decomposition.ipynb]: Fixed in e5c7f6c. Paper-mode reference and invariant checks are recorded as boolean gates rather than asserted. A deliberately mismatched export now completes and retains a failed verdict plus its evidence/provenance artifacts.

File: docs/source/reproductions/na2q-value-decomposition.ipynb [hunk 4]
```diff
@@ -0,0 +1,966 @@
...
+    }
+   ],
+   "source": [
+    "class NA2QDecompositionAdapter(torch.nn.Module):\n",
+    "    def __init__(self, recorded=None):\n",
+    "        super().__init__()\n",
+    "        self.recorded = recorded\n",
+    "\n",
+    "    def forward(self, individual_value, coalition_mask_value, context, identity_semantics):\n",
+    "        if self.recorded is not None:\n",
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/na2q-value-decomposition.ipynb]: Fixed in e5c7f6c by making the boundary explicit: paper mode is an export-mapping audit, `reference_agreement` is not independently evaluated, and `claim_ready` remains false. A deliberately mismatched export was exercised and retained as `reference_export_reconstruction: failed`.

File: docs/source/reproductions/na2q-value-decomposition.ipynb [hunk 5]
```diff
@@ -0,0 +1,966 @@
...
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "def mean_ci(values):\n",
+    "    values = torch.as_tensor(values, dtype=torch.float).flatten()\n",
```
[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/na2q-value-decomposition.ipynb]: Fixed in e5c7f6c. Paper task arrays must be `(5, 32)`, diagnostics must be `(5,)`, method mappings and metric keys are validated, and negative shape-gate checks execute in the smoke notebook.

[the remark this answers is not in the document]
Comment (Xmaster6y) [reply docs/source/reproductions/na2q-value-decomposition.ipynb]: Fixed in e5c7f6c. Task returns are shape-gated at `(5, 32)`, reduced to five seed means, and only those independent seed means are bootstrapped (`n=5`).
